flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/JoinToMultiJoinRule.java

```diff
@@ -234,11 +236,53 @@ public void onMatch(RelOptRuleCall call) {
                         Pair.left(joinSpecs),
                         projFieldsList,
                         com.google.common.collect.ImmutableMap.copyOf(newJoinFieldRefCountsMap),
-                        RexUtil.composeConjunction(rexBuilder, newPostJoinFilters, true));
+                        RexUtil.composeConjunction(rexBuilder, newPostJoinFilters, true),
+                        newCommonJoinKeys);
 
         call.transformTo(multiJoin);
     }
 
+    /**
+     * Creates a Map {join condition -> field names} which has all the conditions containing common
+     * join key and all the keys equal to it.
+     *
+     * @param origJoin original Join node
+     * @param left left child of the Join node
+     * @return Map {join condition -> field names}
+     */
+    private Map<RexCall, Set<String>> combineJoinKeys(Join origJoin, RelNode left) {
+        Map<RexCall, Set<String>> newCondToFieldsMap = getCondToFieldsMap(origJoin);
+
+        if (canCombine(left, origJoin)) {
+            final MultiJoin multiJoin = (MultiJoin) left;
+            Map<RexCall, Set<String>> origCondToKeys = getCondToFieldsMap(origJoin);
+            Map<RexCall, Set<String>> multiJoinCondToKeys = multiJoin.getConditionsToFieldsMap();
+
+            for (Map.Entry<RexCall, Set<String>> origEntry : origCondToKeys.entrySet()) {
+                boolean intersects = false;
+                Set<String> origKeys = origEntry.getValue();
```

Comment:
Can we rename across the rule?
- origJoin -> newJoin
- origKeys -> newJoinKeys

Something like this. I think it'd be a bit easier to understand